### ISSUE TITLE
CS8602 claims false positive on null reference in local function

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn#41689]
**Version Used**: 16.6.0 Preview 1.0 (29812.224.master)

**Steps to Reproduce**:

```
git clone https://github.com/microsoft/vs-threading.git
cd vs-threading\src
git checkout 218ab77a6126819c3a5ba256b6ee5f19b987672d
Microsoft.VisualStudio.Threading.sln
```

Wait for warnings to be reported in the solution.

**Expected Behavior**:

No CS8602 warnings.

**Actual Behavior**:

5 CS8602 warnings:

4 of them (in ThreadingToolsTests.cs) are around [calls to `ThreadingTools.ApplyChangeOptimistically`](https://github.com/AArnott/vs-threading/blob/218ab77a6126819c3a5ba256b6ee5f19b987672d/src/Microsoft.VisualStudio.Threading.Tests/ThreadingToolsTests.cs#L24-L25), which very oddly considers the implicit generic type argument to be `GenericParameterHelper?` instead of simply `GenericParameterHelper`. There's no nullable annotation anywhere to give the compiler the idea that it should be nullable. But a ` : class?` generic type constraint is applied to the `ApplyChangeOptimistically` method itself, but that shouldn't force callers into a nullable annotation, right?
If I add an explicit `<GenericParameterHelper>` generic type argument to the method invocation, the warnings go away.

The last warning is in `VSTHRD109AvoidAssertInAsyncMethodsCodeFix.cs` [where we dereference a local](https://github.com/AArnott/vs-threading/blob/218ab77a6126819c3a5ba256b6ee5f19b987672d/src/Microsoft.VisualStudio.Threading.Analyzers.CodeFixes/VSTHRD109AvoidAssertInAsyncMethodsCodeFix.cs#L103) from within a local method, where the local is itself defined in the outer method. But the local is declared and assigned only once and to a non-null value. Even the compiler recognizes this at the assignment location. Yet within the local function it thinks the local may be null.

Both of these are false positives.

**Comment by jcouv** [comment]
Any chance you could extract a more self-contained repro? Otherwise, I fear this issue will linger due to overhead of getting into it. We'd much appreciate.

**Comment by AArnott** [comment]
Sure, I was able to whittle it down to this, where again the `syntaxNode` dereference within the local method is called out as faulty.

```cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Threading;
using System.Threading.Tasks;

public class VSTHRD109AvoidAssertInAsyncMethodsCodeFix
{
    public async Task RegisterCodeFixesAsync(CodeFixContext context)
    {
        foreach (var diagnostic in context.Diagnostics)
        {
            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
            var syntaxNode = (ExpressionSyntax)root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);

            Task<Solution?> Fix(string fullyQualifiedMethod, IMethodSymbol? methodSymbol, CancellationToken cancellationToken)
            {
                var assertionStatementToRemove = syntaxNode.FirstAncestorOrSelf<StatementSyntax>();

                return Task.FromResult<Solution?>(context.Document.Project.Solution);
            }
        }
    }
}
```

With a project file of:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Library</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="2.8.2" />
    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="2.9.6" PrivateAssets="all" />
  </ItemGroup>
</Project>
```

**Comment by AArnott** [comment]
The above smaller repro only focused on one of the 5 warnings. The other four warnings are now tracked by #41705

**Comment by RikkiGibson** [comment]
Since multiple issues have been opened against all the warnings mentioned in this issue, I will close this out as a dupe.
